### ISSUE TITLE
[Idea] Auto-Updater

### ISSUE THREAD

**Issue by mayuso** [demilich1/metastone#199]
I have been messing with Java again lately and came up with an idea that I feel could also benefit metastone.

Wouldn't it be interesting to implement an Auto-Updater? It's very simple idea (Implementation not so much):

Every time you execute metastone, it checks if there is a new release. If there is, it downloads it and replaces the old files with the new ones.

As you already know by now I am just an amateur developer without much experience, and I still need to learn about this before implementing it, but I think it would be a great addition to metastone.

What do you think?

**Comment by akoscz** [comment]
Take a look at https://github.com/threerings/getdown It's still under development. 
There's also https://github.com/vinumeris/updatefx but hasent had a commit for over 9 months.
Start playing around with them to see how far you can get with integrating one of these into Metastone.

**Comment by mayuso** [comment]
Many thanks for the links :) I will start messing around with those as soon as I get home from work today. I saw Getdown was used for Puzzle Pirates, I used to play this when I was younger, plus as you said, it seems like UpdateFX is a bit abandoned, so I will start with Getdown.

**Comment by mayuso** [comment]
I have been looking into Getdown for a couple hours and it makes no sense to implement it on metastone. 

I did implement it on another project I work on, just to test, and it's pretty much useless. Sometimes works, sometimes it doesn't, and when it doesn't, it corrupts your files and the user needs to download the project again.

Plus it's not integrated in the project _per se_, you need to provide another .jar that is the one that checks for updates and then executes your actual software.

I won't give up and will check UpdateFX tonight or tomorrow, depending on how much free time I can get.

**Comment by demilich1** [comment]
I implemented an auto-updater for a company application recently. Having a second application which handles the patching is the most common solution, because you cannot patch a running program (at least on Windows), so there is not much you can do if you want to patch the .exe.

**Comment by mayuso** [comment]
So the most common is to have 2 executable files? For example:

* MetaStone 1.1.1.exe
* MetaStoneUpdate.exe

This is just an example, the idea here is that the user always executes `MetaStoneUpdate.exe`, and then this executable checks for updates, closes itself and executes the actual metastone, right?

I didn't know about this design choices, seems interesting, I definitely need to learn more about software solution designs, I'll keep this in mind. Thank you @demilich1

**Comment by demilich1** [comment]
Yes, this is exactly how this works @mayuso . Of course, the next question soon arises: What if the updater.exe contains a bug and needs a patch? Again, it cannot patch itself. So what we did to solve this was to add an update check for the updater in the main application. Basically the updater takes care that the main application is always patched and the main application patches the updater.

All in all this is quite complicated for such a common thing as auto-updating, but I don't know of any simple drop-in solution.

**Comment by mayuso** [comment]
Getdown actually supports this features pretty well, I might go back to it and try a better implementation before moving to anything else.

Updating the updater shouldn't be a huge problem IMHO. We can simply implement Getdown into the main project (I steal your idea, sorry xD). But I don't think this will ever be necessary, unless it's complete garbage. The updater is pretty simple (When talking about Getdown).

**Comment by mayuso** [comment]
I have been trying again and again to implement  Getdown on LoLPing (It's a tiny project I develop for fun, it's a lot smaller than metastone, so I assume it should also be easier to implement an auto-update system on it).

The system seems to work just fine except for a file called `digest.txt`, it is supposed to contain an "ID" (a string of characters that identify a specific version of each file (Similar to how  git identifies commits)). 

For the test, I created one version, uploaded and also saved it on my desktop, then I created another version changing a couple label texts (Something simple just to test) and uploaded it. I executed the "old" version I previously stored on my desktop and it finds the files just fine, but it always says `digest.txt` is invalid. I am using getdown 1.5, which I believe is the latest version, and still can't make it work.

I will move to UpdateFX tomorrow.

**Comment by mayuso** [comment]
I can't remember already. I ended up creating my own implementation of an auto-updater, I was going to push it to github and also use it on metastone but a local company ended up buying it, so I couldn't legally make it Open Source.

I'll try to make another implementation and push it, but I can't promise already, I also said I was going to work on the GUI, and I did, but I don't really have much spare time to polish it, I need money so I need to work, I can't live out of Open Source projects (I hope I could).
